### ISSUE TITLE
Codable forms: parsing non-String values

### ISSUE THREAD

**Issue by svanimpe** [IBM-Swift/Kitura#1261]
## Context and Description

I am trying to replace manual form parsing with Codable forms. However, the behavior when dealing with empty values is not as I would expect it.

### Environment Details

Kitura 2.3
Xcode 9.3

### Steps to Reproduce

Codable type:

```
struct LocationForm: Codable {
    let address: String?
    let city: String?
    let country: String?
    let latitude: Double?
    let longitude: Double?
}
```

Form sent: `address=&city=&country=&latitude=&longitude=`

These are all `<input>`s for which a value was never set.

### Expected vs. Actual Behaviour

- [x] Expected to happen: `latitude` and `longitude` are `nil`.

- [x] Actually happened: An error is thrown:

```
[2018-04-30T08:53:50.270+02:00] [ERROR] [QueryDecoder.swift:145 decodingError()] Could not process field named 'latitude'.
[2018-04-30T08:53:50.273+02:00] [ERROR] [RouterMiddlewareWalker.swift:72 next()] dataCorrupted(Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "latitude", intValue: nil)], debugDescription: "Could not process field named \'latitude\'.", underlyingError: nil))
```

I would expect non-String values to be `nil` if either the value isn't provided, or it cannot be parsed to the correct type. This is in line with code Codable forms are supposed to replace, as in the example on https://developer.ibm.com/swift/2018/04/17/announcing-kitura-2-3/, where `id` should be an Int:

```
router.post("/form") { request, response, next in
    guard let encodedBody = request.body,
        case .urlEncoded(let body) = encodedBody,
        let idString = body["id"],
        let id = Int(idString),
        let name = body["name"]
    else {
        return try response.status(.unprocessableEntity).end()
    }
    let user = User(id: id, name: name)
    // Application logic
    response.status(.created).send(json: user)
    next()
}
```

**Comment by Andrew-Lees11** [comment]
Hello Steven

Thank you for raising this issue and providing so much detail.

When we receive a String such as "address=&city=&country=&latitude=&longitude=" we process the = as being an empty string "". Since longitude and latitude are supposed to be doubles this causes the unprocessableEntity error to be thrown. 

To get a nil value for an optional field using the Query decoder the inputs need to not be there. E.g:
address=&city=&country=

Would set address, city, country to be empty string and latitude, longitude would be nil. 

We would expect you to not send anything if you want to get a nil value and use validation on the web end to ensure it your form matches the Codable object you are trying to decode into. 

Would it be possible for you to change your forms so that instead of sending an empty string for nil they don't send the key at all?

**Comment by svanimpe** [comment]
Hi Andrew,

I am using a simple HTML form with a POST action. The String `address=&city=&country=&latitude=&longitude=` is what is generated when the form is submitted. 

I'm not sure I can remove fields from this string without removing the corresponding `<input>` tags. I don't think the latter is common practice, as most people simply clear an input's value, instead of removing the tags.

FYI: My current workaround is to give these fields a value of 0 when clearing them, that way they do get parsed to a Double. But this may not always work.

For example:

Let's say I have a date picker on my page and my user can optionally pick a date. My page also contains the following tag: `<input type="hidden" name="date">`. This tag is assigned a value when the user picks a date.
To really benefit from codable forms, I would then create a codable struct with a `let date: Date?` property and let Kitura handle the parsing and conversion from String to Date.
However, this won't work because when the user doesn't pick a date, Kitura will interpret `date=` as an empty string and throw an error when trying to parse a date.
My workaround also doesn't work here, as I do not want to set a default date (the date is optional, I don't want to force a default).

Also, as a developer, I never see the string that is sent to the server, so I consider it an implementation detail. I simply create a form, submit it and receive it on the other end. Would it then not make more sense to parse empty fields as `nil`, instead of empty strings and errors for all other types?

**Comment by Andrew-Lees11** [comment]
So we have looked into having the decoder turn the empty string to nil for non strings, and having the result be nil if the decoding fails for any reasons and felt these seems like inconsistent behavior for different types. 

Since HTML Forms treat "" as nil we agree that we should also  treat this as nil when decoding to a Swift type.

What this would mean is that "" is not acceptable for a non optional string and would be nil for an Optional string. In the case of date or double that you suggested "address=&city=&country=&latitude=&longitude=" would give you nil for all the values if they are are optional in your struct.

If someone sent a value that does not decode to the type you specified e.g. double = "hello" Then we will continue to return an error.

I'm currently working on a pull request to implement this change and it would be added in a patch.

Would this suggested solution work for your use case and do you have any comments?

**Comment by svanimpe** [comment]
If I understand you correctly, the behavior will then become:

- Key not included in the form string -> nil
- Key included, but no value specified -> nil
- Value specified, but not of the correct type -> error

If so, +1 from me. That would fit great with my use case. I don't use codable forms with codable handlers, but instead use the `read(as:)` method in a regular handler:

```swift
guard let form = try? request.read(as: SomeForm.self) else {
    response.status(.badRequest)
    return next()
}
```

With the new behavior, this guard will only catch actual errors (a required field is missing, or a value is not of the correct type) whereas with the current behavior, `read(as:)` also throws on empty values.

**Comment by Andrew-Lees11** [comment]
Yes that is the behavior we are looking to implement. A pull request with the change to implement this is [here](https://github.com/IBM-Swift/Kitura/pull/1262).
